```{r setup, include=FALSE}
library(tidyverse)
library(synExtra)
library(here)
library(powerjoin)
library(qs)
library(ggbeeswarm)
library(ggrepel)

synapser::synLogin()
syn <- synExtra::synDownloader(normalizePath("~/data"), .cache = TRUE)
```


```{r}
sprint_all <- syn("syn64942445") %>%
  read_csv()

sprint_repeat_edits_old <- syn("syn64376905") %>%
  read_csv()

liu_repeat_counts <- syn("syn64829593") %>%
  qread()

rm_raw <- syn("syn55256576") %>%
  read_csv()

liu_meta <- syn("syn65596451") %>%
  read_csv()
```

Only include LTRs, LINEs, SINEs

Exclude
* Tx1 (only xenopus)
* Dong (arthropods)
* Penelope (drosophila)
* 5S-Deu-L2 (most likely non-functional)
* RTE-BovB (bovine)
* Gypsy (misclassified ERV)
* RTE-X (nematodes - reptiles, misclassified in humans)
* CR1 (avian)
* Jockey (insects, ancient, misclassified)

```{r}
# repeat_whitelist <- rm_raw %>%
#   distinct(class_family) %>%
#   filter(
#     !str_detect(class_family, coll("?")),
#     !str_detect(class_family, coll("Jockey")),
#     !str_detect(class_family, coll("Tx1")),
#     !str_detect(class_family, coll("Dong")),
#     !str_detect(class_family, coll("Penelope")),
#     !str_detect(class_family, coll("5S-Deu-L2")),
#     !str_detect(class_family, coll("tRNA-Deu")),
#     !str_detect(class_family, coll("tRNA-RTE")),
#     !str_detect(class_family, coll("RTE-BovB")),
#     !str_detect(class_family, coll("Gypsy")),
#     !str_detect(class_family, coll("RTE-X")),
#     !str_detect(class_family, coll("CR1")),
#     str_starts(class_family, coll("LTR")) |
#       str_starts(class_family, coll("SINE")) |
#       str_starts(class_family, coll("LINE"))
#   )

repeat_whitelist <- tibble(
  class_family = c(
    "LINE/L1", "LINE/L2", "LTR", "LTR/ERV1", "LTR/ERVK", "LTR/ERVL",
    "LTR/ERVL-MaLR", "Retroposon/SVA", "SINE/Alu", "SINE/MIR",
    "SINE/tRNA"
  )
)
```

```{r}
sprint_all_agg <- sprint_all %>%
  filter(
    type %in% c("AG", "TC"),
  ) %>%
  dplyr::rename(
    nedited = ad,
    unedited = dp,
    sample_id = run_id
  ) %>%
  group_by(
    class_family, sample_id
  ) %>%
  summarise(
    supporting_reads = sum(supporting_reads),
    nedited = sum(nedited),
    unedited = sum(unedited),
    total = nedited + unedited,
    .groups = "drop"
  )

sprint_all_agg_old <- sprint_repeat_edits_old %>%
  filter(
    edit_type %in% c("AG", "TC"),
  ) %>%
  group_by(
    `repeat`, sample_id
  ) %>%
  summarise(
    supporting_reads = sum(score),
    nedited = sum(nedited),
    unedited = sum(unedited),
    total = nedited + unedited,
    .groups = "drop"
  )
```

```{r}
inner_join(
  sprint_all_agg,
  sprint_all_agg_old,
  by = c("class_family" = "repeat", "sample_id"),
  suffix = c("_new", "_old")
) %>%
  ggplot(
    aes(
      total_new,
      total_old
    )
  ) +
  geom_point() +
  coord_equal()

```



```{r}
count(sprint_all, `repeat`)
count(sprint_all, `family`)
```

```{r}
library(lme4)
library(lmerTest)
library(broom)
library(broom.mixed)
```

## Absolute levels

```{r}
abs_test_data <- sprint_all_agg %>%
  semi_join(
    repeat_whitelist,
    by = "class_family"
  ) %>%
  bind_rows(
    group_by(
      .,
      sample_id
    ) %>%
    summarise(
      across(c(nedited, unedited, total, supporting_reads), sum),
      .groups = "drop"
    ) %>%
    mutate(
      class_family = "all"
    )
  ) %>%
  power_inner_join(
    liu_meta %>%
      select(Run, patient_sample_id, tdp_43_status, size_factor),
    by = c("sample_id" = "Run"),
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  mutate(
    across(
      c(nedited, unedited, total, supporting_reads),
      .fns = list(
        adjusted = \(x) x / size_factor
      )
    )
  )

abs_test_data_long <- abs_test_data %>%
  select(class_family, nedited, unedited, total, supporting_reads, patient_sample_id, tdp_43_status, ends_with("adjusted")) %>%
  pivot_longer(
    cols = -c(class_family, patient_sample_id, tdp_43_status),
    names_to = "type",
    values_to = "value"
  )
```


```{r}
count_type_human_readable <- c(
  nedited_adjusted = "Edited reads",
  unedited_adjusted = "Unedited reads",
  total_adjusted = "Total reads",
  supporting_reads_adjusted = "Score adjusted",
  editing_index = "Editing rate\n(edits / 1000 bp of transcript)"
)
```


```{r}
abs_test_paired_raw <- abs_test_data_long %>%
  semi_join(
    repeat_whitelist,
    by = "class_family"
  ) %>%
  mutate(
    across(value, log10)
  ) %>%
  pivot_wider(
    names_from = `tdp_43_status`,
    values_from = -c(`tdp_43_status`, class_family, patient_sample_id, type)
  ) %>%
  mutate(
    difference = `TDP-43 negative` - `TDP-43 positive`
  ) %>%
  group_by(
    class_family, type
  ) %>%
  summarise(
    res = list(
      possibly(lm)(
        difference ~ 1,
        data = pick(everything())
      )
    ),
    .groups = "drop"
  )

abs_test_paired_res <- abs_test_paired_raw %>%
  mutate(
    across(res, \(x) map(x, tidy))
  ) %>%
  unnest(res) %>%
  group_by(
    type
  ) %>%
  mutate(
    padj = p.adjust(p.value, method = "BH")
  ) %>%
  ungroup()


abs_test_paired_res %>%
  filter(type == "nedited_adjusted") %>%
  arrange(p.value)

abs_test_paired_res %>%
  filter(type == "unedited_adjusted") %>%
  arrange(p.value)

abs_test_paired_res %>%
  filter(type == "supporting_reads_adjusted") %>%
  arrange(p.value)

p <- abs_test_paired_res %>%
  filter(
    type %in% c("nedited_adjusted", "unedited_adjusted", "total_adjusted")
  ) %>%
  ggplot(
    aes(
      estimate,
      -log10(p.value)
    )
  ) +
  geom_point() +
  facet_wrap(~type, scales = "free_y")

p <- abs_test_paired_res %>%
  filter(
    type %in% c("nedited_adjusted", "unedited_adjusted", "total_adjusted")
  ) %>%
  ggplot(
    aes(p.value)
  ) +
  geom_histogram() +
  facet_wrap(vars(type), scales = "free")


library(lme4)
library(lmerTest)
abs_test_random_raw <- abs_test_data_long %>%
  crossing(
    trans = c("log10", "linear")
  ) %>%
  mutate(
    value = if_else(trans == "log10", log10(value), value)
  ) %>%
  group_by(
    type, `repeat`, trans
  ) %>%
  summarise(
    res = list(
      possibly(lmer)(
        value ~ tdp_43_status + (1 | patient_sample_id),
        data = cur_data()
      )
    ),
    .groups = "drop"
  )

abs_test_random_raw %>%
  filter(trans == "log10", `repeat` == "LTR/ERVK", type == "nedited_adjusted") %>%
  chuck("res", 1) %>%
  summary()

abs_test_paired_res %>%
  filter(type == "nedited_adjusted", trans == "log10", `repeat` == "LTR/ERVK")

abs_test_res <- abs_test_raw %>%
  mutate(
    across(res, \(x) map(x, tidy))
  ) %>%
  unnest(res)
s
```

```{r}
patient_id_color_mapping <- liu_meta %>%
  filter(
    tdp_43_status == "TDP-43 positive"
  ) %>%
  distinct(patient_sample_id) %>%
  arrange() %>%
  pull(patient_sample_id) %>% {
    set_names(
      paletteer::paletteer_d("colorblindr::OkabeIto", length(.)),
      .
    )
  }


ps <- abs_test_data_long %>%
  filter(type %in% c("nedited_adjusted", "unedited_adjusted", "total_adjusted", "supporting_reads_adjusted")) %>%
  semi_join(
    repeat_whitelist,
    by = "class_family"
  ) %>%
  group_by(
    type
  ) %>%
  summarize(
    p = list({
      cur_type <- cur_group()$type
      ggplot(
        pick(everything()),
        aes(
          tdp_43_status,
          value,
          color = patient_sample_id
        )
      ) +
      geom_line(
        aes(group = patient_sample_id)
      ) +
      geom_quasirandom(width = .1) +
      scale_color_manual(
        values = patient_id_color_mapping,
        guide = "none"
      ) +
      ggnewscale::new_scale_color() +
      geom_text(
        aes(label = signif(padj, 2), color = padj < 0.05),
        x = -Inf, y = Inf,
        hjust = 0, vjust = 1,
        data = filter(
          abs_test_paired_res,
          type == cur_type
        ),
        inherit.aes = FALSE
      ) +
      scale_color_manual(
        values = c(`TRUE` = "red", `FALSE` = "black"),
        guide = "none"
      ) +
      scale_y_log10(
        expand = expansion(mult = c(.1, .2))
      ) +
      coord_cartesian(
        clip = "off"
      ) +
      ggh4x::facet_wrap2(~class_family, scales = "free_y", axes = "y") +
      theme(
        axis.text.x = element_text(angle = 45, hjust = 1)
      ) +
      labs(
        x = NULL,
        y = count_type_human_readable[cur_type],
        title = count_type_human_readable[cur_type]
      )
    })
  )

dir.create(
  here("dsrna", "plots", "sprint_analysis"),
  showWarnings = FALSE
)
pwalk(
  ps,
  \(p, type, ...)
    ggsave(
      here("dsrna", "plots", "sprint_analysis", paste0("sprint_quants_scatter_selected_", type, ".pdf")),
      p, width = 6, height = 5
    )
)
```

## Proportion test

## Agg

```{r}
prop_test_repeat_agg <- sprint_all_agg %>%
  group_nest(
    `repeat`
  ) %>%
  mutate(
    mod = map2(
      data, `repeat`,
      \(x, y) {
        message(y)
        r <- safely(glm)(
          cbind(nedited, unedited) ~ pathology,
          data = x,
          family = binomial
        )
        message(as.character(r$error))
        r
      }
    )
  )

prop_test_repeat_agg_res <- prop_test_repeat_agg %>%
  mutate(
    res = map(mod, \(x) if (!is.null(x$result)) tidy(x$result) else NULL)
  ) %>%
  select(-data, -mod) %>%
  unnest(res)

```

```{r}
p <- prop_test_repeat_agg_res %>%
  filter(
    term == "pathologydisease",
    !str_ends(`repeat`, fixed("?"))
  ) %>%
  arrange(estimate) %>%
  mutate(
    `repeat` = fct_inorder(`repeat`)
  ) %>%
  ggplot(
    aes(
      estimate,
      `repeat`
    )
  ) +
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_errorbarh(
    aes(xmin = estimate - std.error, xmax = estimate + std.error)
  ) +
  geom_point(
    aes(fill = p.value < 0.05),
    shape = 21,
    color = "black",
  ) +
  scale_fill_manual(
    values = c(`TRUE` = "black", `FALSE` = "grey")
  ) +
  labs(
     x = "Log odds ratio edit rate",
     y = NULL
  )

dir.create("sprint_analysis", showWarnings = FALSE)
ggsave(
  "sprint_analysis/prop_test_glmm_repeat.pdf",
  p,
  width = 8, height = 6
)

```


## Individual

```{r}

prop_test_repeat <- sprint_all %>%
  group_nest(
    `repeat`
  ) %>%
  filter(
    map_lgl(data, \(x) nrow(x) > 10)
  ) %>%
  mutate(
    mod = map2(
      data, `repeat`,
      \(x, y) {
        message(y)
        r <- safely(glmer)(
          cbind(nedited, unedited) ~ 1 + pathology + (1 | repeat_id),
          data = x,
          family = binomial,
          verbose = 1
        )
        message(as.character(r$error))
        r
      }
    )
  )

qsave(
  prop_test_repeat,
  "prop_test_repeat_raw.qs"
)
```


```{r}
prop_test_repeat_res <- prop_test_repeat %>%
  mutate(
    res = map(mod, \(x) if (!is.null(x$result)) tidy(x$result) else NULL)
  ) %>%
  select(-data, -mod) %>%
  unnest(res)


prop_test_repeat_res %>%
  filter(term == "pathologydisease") %>%
  View()
```

```{r}
p <- prop_test_repeat_res %>%
  filter(
    term == "pathologydisease",
    !str_ends(`repeat`, fixed("?"))
  ) %>%
  arrange(estimate) %>%
  mutate(
    `repeat` = fct_inorder(`repeat`)
  ) %>%
  ggplot(
    aes(
      estimate,
      `repeat`
    )
  ) +
  geom_vline(xintercept = 0, linetype = "dashed") +
  geom_errorbarh(
    aes(xmin = estimate - std.error, xmax = estimate + std.error)
  ) +
  geom_point(
    aes(fill = p.value < 0.05),
    shape = 21,
    color = "black",
  ) +
  scale_fill_manual(
    values = c(`TRUE` = "black", `FALSE` = "grey")
  ) +
  labs(
     x = "Log odds ratio edit rate",
     y = NULL
  )

dir.create("sprint_analysis", showWarnings = FALSE)
ggsave(
  "sprint_analysis/prop_test_glmm_repeat.pdf",
  p,
  width = 8, height = 6
)
 ```

## Compare agg vs individual

```{r}
prop_test_repeat_comp <- bind_rows(
  agg = prop_test_repeat_agg_res,
  ind = prop_test_repeat_res,
  .id = "type"
)

p <- prop_test_repeat_comp %>%
  filter(
    term == "pathologydisease",
    !str_ends(`repeat`, fixed("?"))
  ) %>%
  select(type, `repeat`, estimate, std.error, p.value) %>%
  pivot_wider(
    names_from = type,
    values_from = c(estimate, std.error, p.value)
  ) %>%
  ggplot(
    aes(
      estimate_agg,
      estimate_ind,
      text = `repeat`
    )
  ) +
  geom_abline() +
  geom_point()

plotly::ggplotly(p)
```


```{r}
prop_test_family <- sprint_all %>%
  group_nest(
    family
  ) %>%
  filter(
    map_lgl(data, \(x) nrow(x) > 10)
  ) %>%
  mutate(
    mod = map2(
      data, family,
      \(x, y) {
        message(y)
        r <- safely(glmer)(
          cbind(nedited, unedited) ~ pathology + (1 | repeat_id) + (1 | sample_id),
          data = x,
          family = binomial
        )
        message(as.character(r$error))
        r
      }
    )
  )

```



## Quantifying total reads on repeats

1. Check overall number of editing events TDP-43 vs Control
2. Compare absolute numbers
3. Call highly edited sites with clusters of edits, do they lose editing in TDP-43?
4. Check length of reads in original fastq files. Do we get reads shorter than 250bp?
5. Check top genes enriched in C9 from Chan core in my analysis



```{r}
library(DESeq2)

rm_whitelisted <- rm_raw %>%
  mutate(
    gene_id = paste(rep_name, row_number(), sep = "_")
  ) %>%
  power_inner_join(
    repeat_whitelist,
    by = "class_family",
    check = check_specs(
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn"
    )
  )

repeat_counts_mat <- liu_repeat_counts[["counts"]] %>%
  magrittr::set_colnames(
    str_remove(colnames(.), fixed(".sorted.bam"))
  ) %>% {
    .[rm_whitelisted$gene_id, liu_meta$Run]
  }

de_meta <- liu_meta %>%
  select(
    Run, patient_sample_id, tdp_43_status
  )

de <- DESeqDataSetFromMatrix(
  repeat_counts_mat,
  de_meta,
  design = ~ tdp_43_status + patient_sample_id
)
sizeFactors(de) <- liu_meta$size_factor
des <- DESeq(de)

qsave(
  des,
  file.path("dsrna", "data", "liu_repeat_de.qs")
)
des <- qread(file.path("dsrna", "data", "liu_repeat_de.qs"))
```


```{r}
repeat_all_counts <- list(
  normalized = counts(des, normalized = TRUE),
  raw = counts(des, normalized = FALSE)
)

repeat_norm_counts_family <- repeat_all_counts %>%
  map(\(x) as_tibble(x, rownames = "gene_id")) %>%
  bind_rows(.id = "normalized") %>%
  pivot_longer(
    -c(normalized, gene_id),
    names_to = "sample_id",
    values_to = "count"
  ) %>%
  pivot_wider(
    names_from = normalized,
    values_from = count
  ) %>%
  dplyr::rename(
    count = normalized,
    raw_count = raw
  ) %>%
  power_inner_join(
    rm_raw %>%
      transmute(
        gene_id = paste(rep_name, row_number(), sep = "_"),
        class_family,
        chromosome, start, end
      ),
    by = "gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )

repeat_norm_counts_family_agg <- repeat_norm_counts_family %>%
  group_by(
    sample_id, class_family
  ) %>%
  summarize(
    count = sum(count),
    raw_count = sum(raw_count),
    .groups = "drop"
  ) %>%
  semi_join(
    repeat_whitelist,
    by = "class_family"
  ) %>%
  power_inner_join(
    liu_meta %>%
      select(Run, patient_sample_id, tdp_43_status),
    by = c("sample_id" = "Run"),
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn"
    )
  )
```


```{r}
repeat_norm_counts_family_agg_long <- repeat_norm_counts_family_agg %>%
  select(class_family, count, patient_sample_id, tdp_43_status) %>%
  pivot_wider(
    names_from = tdp_43_status,
    values_from = count
  ) %>%
  mutate(
    difference = `TDP-43 negative` - `TDP-43 positive`,
    lratio = log2(`TDP-43 negative` / `TDP-43 positive`)
  )

de_agg_raw <- repeat_norm_counts_family_agg_long %>%
  group_by(
    class_family
  ) %>%
  summarize(
    res = list(
      lm(
        difference ~ 1,
        data = pick(everything())
      ) %>%
        broom::tidy()
    ),
    .groups = "drop"
  ) %>%
  unnest(res) %>%
  mutate(
    padj = p.adjust(p.value, method = "BH")
  )


de_agg_raw_lratio <- repeat_norm_counts_family_agg_long %>%
  group_by(
    class_family
  ) %>%
  summarize(
    res = list(
      lm(
        lratio ~ 1,
        data = pick(everything())
      ) %>%
        broom::tidy()
    ),
    .groups = "drop"
  ) %>%
  unnest(res) %>%
  mutate(
    padj = p.adjust(p.value, method = "BH"),
    estimate_fc = 2^estimate
  )
```

```{r}
p <- de_agg_raw_lratio %>%
  # arrange(estimate) %>%
  mutate(
    across(
      class_family,
      \(x) fct_reorder(x, estimate)
    )
  ) %>%
  ggplot(
    aes(
      estimate,
      class_family,
      fill = padj < 0.05
    )
  ) +
  geom_col() +
  geom_text(
    aes(label = signif(padj, 2)),
    hjust = 1,
    nudge_x = -.001,
    color = "white"
  ) +
  scale_fill_manual(
    values = c(`TRUE` = "#c62a2a", `FALSE` = "#6d6868"),
    guide = "none"
  ) +
  labs(
    x = "log2 fold-change TDP-43 (+) vs (-)",
    y = NULL
  )


p
```


```{r}
p <- ggplot(
  repeat_norm_counts_family_agg,
  aes(
    tdp_43_status,
    count,
    color = patient_sample_id
  )
) +
  geom_line(
    aes(group = patient_sample_id)
  ) +
  geom_quasirandom(width = .1) +
  scale_color_manual(
    values = patient_id_color_mapping,
    guide = "none"
  ) +
  ggnewscale::new_scale_color() +
  geom_text(
    aes(label = signif(padj, 2), color = padj < 0.05),
    x = -Inf, y = Inf,
    hjust = 0, vjust = 1,
    data = de_agg_raw_lratio,
    inherit.aes = FALSE
  ) +
  scale_color_manual(
    values = c(`TRUE` = "red", `FALSE` = "black"),
    guide = "none"
  ) +
  scale_y_log10(
    expand = expansion(mult = c(.1, .2))
  ) +
  ggh4x::facet_wrap2(~class_family, scales = "free_y", axes = "y") +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1)
  ) +
  coord_cartesian(
    clip = "off"
  ) +
  labs(
    x = NULL,
    y = "Aggregated count",
    title = "Read count on entire repeat body"
  )
p

ggsave(
  file.path("dsrna", "plots", "repeat_count_by_patient.pdf"),
  p, width = 6, height = 5
)
```


```{r}
repeat_norm_counts_family_agg_scaled <- repeat_norm_counts_family_agg %>%
  group_by(
    class_family
  ) %>%
  mutate(
    count_scaled = scale(count)[, 1]
  ) %>%
  ungroup()

p <- repeat_norm_counts_family_agg_scaled %>%
  ggplot(
    aes(
      class_family,
      count_scaled,
      color = patient_sample_id
    )
  ) +
  geom_line(
    aes(group = patient_sample_id)
  ) +
  geom_point() +
  facet_wrap(
    vars(tdp_43_status)
  ) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1)
  ) +
  labs(
    x = NULL,
    y = "Scaled count",
    color = "Patient"
  )

p
ggsave(
  file.path("dsrna", "plots", "repeat_count_scaled_by_patient.pdf"),
  p, width = 7, height = 4
)


p <- repeat_norm_counts_family_agg_scaled %>%
  ggplot(
    aes(
      class_family,
      count_scaled,
      color = patient_sample_id,
      linestyle = tdp_43_status
    )
  ) +
  geom_line(
  ) +
  geom_point() +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1)
  ) +
  labs(
    x = NULL,
    y = "Scaled count",
    color = "Patient"
  )
p
```


```{r}
extract_deseq_result <- function(de, contrast, name) {
  if (missing(contrast)) {
    res <- results(de, name = name)
  } else {
    res <- results(de, contrast = contrast)
  }
  shrunken <- lfcShrink(de, res = res, type = "ashr")
  shrunken %>%
    as_tibble(rownames = "gene_id") %>%
    left_join(
      res %>%
        as_tibble(rownames = "gene_id") %>%
        select(gene_id, log2FoldChange, lfcSE),
      by = "gene_id", suffix = c("", "_MLE")
    )
}
```

```{r}
resultsNames(des)
res_tdp43 <- extract_deseq_result(
  des,
  name = "tdp_43_status_TDP.43.positive_vs_TDP.43.negative"
)

```


```{r}
p <- res_tdp43 %>%
  ggplot(
    aes(
      log2FoldChange,
      -log10(pvalue),
      color = padj < 0.05
    )
  ) +
  geom_point()

ggsave(
  file.path("dsrna", "plots", "repeat_deseq_tdp43_volcano.png"),
  p, width = 6, height = 6
)
```


```{r}
res_tdp43_family <- res_tdp43 %>%
  power_inner_join(
    rm_raw %>%
      transmute(
        gene_id = paste(rep_name, row_number(), sep = "_"),
        class_family,
        chromosome, start, end
      ),
    by = "gene_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  )

write_csv(
  res_tdp43_family,
  file.path("dsrna", "data", "repeat_tdp43_pos_vs_neg_deseq.csv.gz")
)
# res_tdp43_family <- read_csv(file.path("dsrna", "data", "repeat_tdp43_pos_vs_neg_deseq.csv.gz"))
```


```{r}
p <- res_tdp43_family %>%
  semi_join(
    repeat_whitelist,
    by = "class_family"
  ) %>%
  arrange(desc(pvalue)) %>%
  ggplot(
    aes(
      log2FoldChange,
      -log10(pvalue),
      color = padj < 0.05
    )
  ) +
  ggrastr::rasterize(
    geom_point(shape = 16, alpha = .6),
    dpi = 200, dev = "ragg"
  ) +
  scale_color_manual(
    values = c(`TRUE` = "red", `FALSE` = "black")
  ) +
  coord_cartesian(
    xlim = c(-6, 5)
  ) +
  facet_wrap(
    vars(class_family)
  )

ggsave(
  file.path("dsrna", "plots", "repeat_deseq_tdp43_volcano_by_family_selectd.pdf"),
  p, width = 7, height = 5
)
```

```{r}
res_tdp43_family_stats <- res_tdp43_family %>%
  filter(
    class_family != ".",
    !str_ends(class_family, fixed("?"))
  ) %>%
  group_by(
    class_family
  ) %>%
  summarize(
    n = n(),
    n_de = sum(padj < .05, na.rm = TRUE),
    n_down = sum(log2FoldChange < 0 & padj < .05, na.rm = TRUE),
    n_up = sum(log2FoldChange > 0 & padj < .05, na.rm = TRUE),
    prop_de = n_de / n,
    prop_down = n_down / n,
    prop_up = n_up / n,
    .groups = "drop"
  )

highlighted_rep_family <- c(
  "LINE/L1", "LINE/L2",
  "SINE/Alu",
  "LTR/ERV1", "LTR/ERVK", "LTR/ERVL",
  "DNA/TcMar-Tigger",
  "Satellite/centr"
)

p <- res_tdp43_family_stats %>%
  ggplot(
    aes(
      n_de,
      prop_de,
      text = class_family
    )
  ) +
  geom_point() +
  geom_text_repel(
    aes(
      label = class_family
    ),
    data = \(x) mutate(
      x,
      class_family = if_else(
        class_family %in% highlighted_rep_family,
        class_family,
        ""
      )
    ),
    force = 5,
    box.padding = 1,
    point.padding = 5e-6,
    max.overlaps = Inf,
    max.iter = 1e4,
    seed = 42,
    min.segment.length = 0
  ) +
  scale_x_log10() +
  labs(
    x = "N differentially expressed",
    y = "Proportion differentially expressed"
  )

plotly::ggplotly(
  p
)

ggsave(
  file.path("dsrna", "plots", "repeat_deseq_tdp43_de_vs_prop.pdf"),
  p, width = 6, height = 4
)


p <- res_tdp43_family_stats %>%
  ggplot(
    aes(
      n_up,
      prop_up,
      text = class_family
    )
  ) +
  geom_point() +
  geom_text_repel(
    aes(
      label = class_family
    ),
    data = \(x) mutate(
      x,
      class_family = if_else(
        class_family %in% highlighted_rep_family,
        class_family,
        ""
      )
    ),
    force = 5,
    box.padding = 1,
    point.padding = 5e-6,
    max.overlaps = Inf,
    max.iter = 1e4,
    seed = 42,
    min.segment.length = 0
  ) +
  scale_x_log10() +
  labs(
    x = "N differentially expressed",
    y = "Proportion differentially expressed"
  )

plotly::ggplotly(
  p
)

```



```{r}
res_tdp43_family %>%
  filter(
    class_family != ".",
    !str_ends(class_family, fixed("?"))
  )

```


```{r}
library(ggridges)

p <- res_tdp43_family %>%
  filter(
    class_family != ".",
    !str_ends(class_family, fixed("?"))
  ) %>%
  ggplot(
    aes(
      x = log2FoldChange,
      y = class_family
    )
  ) +
  geom_density_ridges() +
  geom_vline(xintercept = 0, linetype = "dashed") +
  coord_cartesian(xlim = c(-2, 2))
p
```

```{r}
res_tdp43_family_ttest <- res_tdp43_family %>%
  filter(
    class_family != ".",
    !str_ends(class_family, fixed("?"))
  ) %>%
  group_by(
    class_family
  ) %>%
  summarize(
    res = t.test(
      log2FoldChange,
      mu = 0
    ) %>%
      broom::tidy() %>%
      list(),
    .groups = "drop"
  ) %>%
  unnest(res) %>%
  mutate(
    padj = p.adjust(p.value, method = "BH")
  )
```


```{r}
sprint_all_agg_normalized <- sprint_all_agg %>%
  power_inner_join(
    liu_meta %>%
      select(sample_id = Run, size_factor),
    by = "sample_id",
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  mutate(
    across(
      c(nedited, unedited, total, supporting_reads),
      .fns = list(
        normalized = \(x) x / size_factor
      )
    )
  )

editing_index_df <- repeat_norm_counts_family_agg %>%
  semi_join(
    repeat_whitelist,
    by = "class_family"
  ) %>%
  power_inner_join(
    abs_test_data,
    by = c("sample_id", "class_family", "patient_sample_id", "tdp_43_status"),
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn",
      duplicate_keys_left = "warn"
    )
  ) %>%
  mutate(
    editing_index = 1000 * nedited_adjusted / count
  )

write_csv(
  editing_index_df,
  here("dsrna", "data", "editing_index.csv.gz")
)
```


```{r}
editing_index_paired_difference <- editing_index_df %>%
  transmute(
    class_family,
    patient_sample_id,
    tdp_43_status,
    count,
    nedited_adjusted,
    editing_index = editing_index
  ) %>%
  pivot_wider(
    names_from = tdp_43_status,
    values_from = c(editing_index, count, nedited_adjusted)
  ) %>%
  mutate(
    difference = `count_TDP-43 negative` - `count_TDP-43 positive`,
    log_diff = log2(`count_TDP-43 negative` / `count_TDP-43 positive`)
  )

editing_index_paired_res <- editing_index_paired_difference %>%
  drop_na(difference) %>%
  group_by(
    class_family
  ) %>%
  filter(
    n() == 7
  ) %>%
  summarize(
    res = list(
      possibly(lm)(
        log_diff ~ 1,
        data = pick(everything())
      ) %>%
        broom::tidy()
    ),
    .groups = "drop"
  ) %>%
  unnest(res) %>%
  mutate(
    padj = p.adjust(p.value, method = "BH")
  )

library(lme4)
editing_index_paired_glmer_res <- editing_index_df %>%
  mutate(
    across(c(count, nedited_adjusted), round)
  ) %>%
  group_by(
    class_family
  ) %>%
  summarize(
    res = list(
      possibly(glmer)(
        cbind(nedited_adjusted, count) ~ tdp_43_status + (1 | patient_sample_id),
        data = pick(everything()),
        family = binomial
      ) %>%
        broom::tidy()
    ),
    .groups = "drop"
  ) %>%
  unnest(res) %>%
  filter(term == "tdp_43_statusTDP-43 positive") %>%
  mutate(
    padj = p.adjust(p.value, method = "BH")
  )


library(glmmTMB)
editing_index_paired_nb_raw <- editing_index_df %>%
  # mutate(
  #   across(c(count, nedited_adjusted), round)
  # ) %>%
  mutate(
    tdp_43_status = factor(tdp_43_status, levels = c("TDP-43 positive", "TDP-43 negative"))
  ) %>%
  group_by(
    class_family
  ) %>%
  summarize(
    res = list(
      possibly(glmmTMB)(
        nedited ~ tdp_43_status + offset(log(raw_count / size_factor)) + (1 | patient_sample_id),
        data = pick(everything()),
        family = nbinom2
      )
    ),
    .groups = "drop"
  )

editing_index_paired_nb_res <- editing_index_paired_nb_raw %>%
  mutate(
    res = map(res, tidy)
  ) %>%
  unnest(res) %>%
  filter(term == "tdp_43_statusTDP-43 negative") %>%
  mutate(
    padj = p.adjust(p.value, method = "BH")
  )
```


```{r}
p <- editing_index_df %>%
  ggplot(
    aes(
      tdp_43_status,
      editing_index,
      color = patient_sample_id
    )
  ) +
  geom_line(
    aes(group = patient_sample_id)
  ) +
  geom_quasirandom(width = .1) +
  scale_color_manual(
    values = patient_id_color_mapping,
    guide = "none"
  ) +
  ggnewscale::new_scale_color() +
  geom_text(
    aes(label = signif(padj, 2), color = padj < 0.05),
    x = -Inf, y = Inf,
    hjust = 0, vjust = 1,
    data = editing_index_paired_nb_res,
    inherit.aes = FALSE
  ) +
  scale_color_manual(values = c(`TRUE` = "red", `FALSE` = "black"), guide = "none") +
  scale_y_log10(
    expand = expansion(mult = c(.1, .2))
  ) +
  ggh4x::facet_wrap2(~class_family, scales = "free_y", axes = "y") +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 1)
  ) +
  labs(
    x = NULL,
    y = "Editing index"
  )

ggsave(
  file.path("dsrna", "plots", "sprint_analysis", "editing_index_paired_nb.pdf"),
  p, width = 6, height = 5
)
```


### Scatter plot editing index difference vs count difference


```{r}
editing_index_vs_count_diff <- power_inner_join(
  de_agg_raw_lratio,
  editing_index_paired_nb_res,
  by = "class_family",
  check = check_specs(
    unmatched_keys_left = "warn",
    unmatched_keys_right = "warn",
    duplicate_keys_right = "warn",
    duplicate_keys_left = "warn"
  ),
  suffix = c("_de", "_ei")
)

editing_index_vs_count_diff_long <- bind_rows(
  `Total count` = de_agg_raw_lratio,
  `Editing index` = editing_index_paired_nb_res,
  .id = "type"
) %>%
  mutate(
    padj_bins = cut(
      padj,
      breaks = c(0, .01, .05, 1),
      labels = c("<0.01", "<0.05", ">=0.05")
    )
  )

editing_index_vs_count_diff_dot_plot <- ggplot(
  editing_index_vs_count_diff_long,
  aes(
    x = type,
    y = class_family
  )
) +
  geom_point(
    aes(
      color = estimate,
      size = padj_bins
    ),
    shape = 15
  ) +
  scale_size_manual(
    values = c("<0.01" = 8, "<0.05" = 4, ">=0.05" = 2)
  ) +
  scale_color_distiller(
    palette = "RdBu",
    limits = c(-.5, .5),
    oob = scales::squish
  ) +
  scale_x_discrete(
    position = "top",
    # expand = expansion(mult = c(.15, .15))
  ) +
  labs(
    x = NULL,
    y = NULL,
    color = "log2\nfold-change",
    size = "Adjusted\np-value"
  ) +
  theme(
    axis.text.x = element_text(angle = 45, hjust = 0, vjust = 0)
  )
editing_index_vs_count_diff_dot_plot

ggsave(
  file.path("dsrna", "plots", "sprint_analysis", "editing_index_vs_count_diff_dot_plot.pdf"),
  editing_index_vs_count_diff_dot_plot, width = 3, height = 4
)
```


Per patient editing index vs absolute difference


```{r}
per_sample_editing_vs_absolute <- power_inner_join(
  repeat_norm_counts_family_agg_long,
  editing_index_df %>%
    select(
      patient_sample_id, class_family, tdp_43_status, editing_index
    ) %>%
    pivot_wider(
      names_from = tdp_43_status,
      values_from = editing_index
    ) %>%
    mutate(
      difference = `TDP-43 negative` - `TDP-43 positive`,
      lratio = log2(`TDP-43 negative` / `TDP-43 positive`)
    ),
  by = c("patient_sample_id", "class_family"),
  suffix = c("_count", "_ei"),
  check = check_specs(
    unmatched_keys_left = "warn",
    unmatched_keys_right = "warn",
    duplicate_keys_right = "warn",
    duplicate_keys_left = "warn"
  )
)

p <- ggplot(
  per_sample_editing_vs_absolute %>%
    filter(
      class_family != "SINE/tRNA"
    ),
  aes(
    x = lratio_count,
    y = difference_ei,
    color = patient_sample_id
  )
) +
  geom_point() +
  scale_color_manual(
    values = patient_id_color_mapping,
    guide = "none"
  ) +
  ggh4x::facet_wrap2(
    vars(class_family),
    scales = "free",
    axes = "all"
  )
p
```


## Compute editing index separately for differentially expressed TEs

```{r}
upregulated_tes <- res_tdp43_family %>%
  filter(
    log2FoldChange < 0,
    padj < .05
  )

n_up_control_tes <- res_tdp43_family %>%
  mutate(
    upregulated = if_else(
      log2FoldChange < 0 & padj < .05,
      "Upregulated", "Control"
    ) %>%
      replace_na("Control")
  ) %>%
  dplyr::count(class_family,upregulated)

repeat_norm_counts_family_agg_by_de <- repeat_norm_counts_family %>%
  semi_join(
    repeat_whitelist,
    by = "class_family"
  ) %>%
  power_left_join(
    upregulated_tes %>%
      transmute(gene_id, upregulated = TRUE),
    by = "gene_id",
    check = check_specs(
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  mutate(across(upregulated, \(x) replace_na(x, FALSE))) %>%
  group_by(
    sample_id, class_family, upregulated
  ) %>%
  summarize(
    count = sum(count),
    raw_count = sum(raw_count),
    .groups = "drop"
  ) %>%
  power_inner_join(
    liu_meta %>%
      select(Run, patient_sample_id, tdp_43_status),
    by = c("sample_id" = "Run"),
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn"
    )
  )



sprint_all_agg_de <- sprint_all %>%
  semi_join(
    repeat_whitelist,
    by = "class_family"
  ) %>%
  power_left_join(
    upregulated_tes %>%
      transmute(gene_id, upregulated = TRUE),
    by = c("gene_id"),
    check = check_specs(
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  mutate(across(upregulated, \(x) replace_na(x, FALSE))) %>%
  filter(
    type %in% c("AG", "TC"),
  ) %>%
  dplyr::rename(
    nedited = ad,
    total = dp,
    sample_id = run_id
  ) %>%
  group_by(
    class_family, sample_id, upregulated
  ) %>%
  summarise(
    supporting_reads = sum(supporting_reads),
    nedited = sum(nedited),
    total = sum(total),
    .groups = "drop"
  )

sprint_all_agg_de_norm <- sprint_all_agg_de %>%
  semi_join(
    repeat_whitelist,
    by = "class_family"
  ) %>%
  bind_rows(
    group_by(
      .,
      sample_id
    ) %>%
    summarise(
      across(c(nedited, total, supporting_reads), sum),
      .groups = "drop"
    ) %>%
    mutate(
      class_family = "all"
    )
  ) %>%
  power_inner_join(
    liu_meta %>%
      select(Run, patient_sample_id, tdp_43_status, size_factor),
    by = c("sample_id" = "Run"),
    check = check_specs(
      unmatched_keys_left = "warn",
      duplicate_keys_right = "warn"
    )
  ) %>%
  mutate(
    across(
      c(nedited, total, supporting_reads),
      .fns = list(
        adjusted = \(x) x / size_factor
      )
    )
  )



editing_index_de_df <- repeat_norm_counts_family_agg_by_de %>%
  semi_join(
    repeat_whitelist,
    by = "class_family"
  ) %>%
  power_inner_join(
    sprint_all_agg_de_norm,
    by = c("sample_id", "class_family", "patient_sample_id", "tdp_43_status", "upregulated"),
    check = check_specs(
      unmatched_keys_left = "warn",
      unmatched_keys_right = "warn",
      duplicate_keys_right = "warn",
      duplicate_keys_left = "warn"
    )
  ) %>%
  mutate(
    editing_index = 1000 * nedited_adjusted / count
  )



library(glmmTMB)
editing_index_de_paired_nb_raw <- editing_index_de_df %>%
  # mutate(
  #   across(c(count, nedited_adjusted), round)
  # ) %>%
  mutate(
    tdp_43_status = factor(tdp_43_status, levels = c("TDP-43 positive", "TDP-43 negative"))
  ) %>%
  group_by(
    class_family, upregulated
  ) %>%
  summarize(
    res = list(
      possibly(glmmTMB)(
        nedited ~ tdp_43_status + offset(log(raw_count / size_factor)) + (1 | patient_sample_id),
        data = pick(everything()),
        family = nbinom2
      )
    ),
    .groups = "drop"
  )

editing_index_de_paired_nb_res <- editing_index_de_paired_nb_raw %>%
  mutate(
    res = map(res, broom::tidy)
  ) %>%
  unnest(res) %>%
  filter(term == "tdp_43_statusTDP-43 negative") %>%
  mutate(
    padj = p.adjust(p.value, method = "BH")
  )


editing_index_de_by_upregulated_raw <- editing_index_de_df %>%
  # mutate(
  #   across(c(count, nedited_adjusted), round)
  # ) %>%
  mutate(
    tdp_43_status = factor(tdp_43_status, levels = c("TDP-43 positive", "TDP-43 negative"))
  ) %>%
  group_by(
    class_family, tdp_43_status
  ) %>%
  summarize(
    res = list(
      possibly(glmmTMB)(
        nedited ~ upregulated + offset(log(raw_count / size_factor)) + (1 | patient_sample_id),
        data = pick(everything()),
        family = nbinom2
      )
    ),
    .groups = "drop"
  )

editing_index_de_by_upregulated_res <- editing_index_de_by_upregulated_raw %>%
  mutate(
    res = map(res, broom::tidy)
  ) %>%
  unnest(res) %>%
  filter(term == "upregulatedTRUE") %>%
  mutate(
    padj = p.adjust(p.value, method = "BH")
  )


library(glmmTMB)
editing_index_de_combined_nb_raw <- editing_index_de_df %>%
  # mutate(
  #   across(c(count, nedited_adjusted), round)
  # ) %>%
  mutate(
    tdp_43_status = factor(tdp_43_status, levels = c("TDP-43 positive", "TDP-43 negative"))
  ) %>%
  group_by(
    class_family
  ) %>%
  summarize(
    res = list(
      possibly(glmmTMB)(
        nedited ~ tdp_43_status*upregulated + offset(log(raw_count / size_factor)) + (1 | patient_sample_id),
        data = pick(everything()),
        family = nbinom2
      )
    ),
    .groups = "drop"
  )

editing_index_de_combined_nb_res <- editing_index_de_combined_nb_raw %>%
  mutate(
    res = map(res, broom::tidy)
  ) %>%
  unnest(res) %>%
  filter(!str_detect(term, "Intercept")) %>%
  mutate(
    padj = p.adjust(p.value, method = "BH")
  )

```

### Plotting editing index for upregulated TEs

```{r}
library(glue)
library(ggtext)

ps <- editing_index_de_df %>%
  mutate(
    upregulated = if_else(upregulated, "Upregulated", "Control") %>%
      factor(levels = c("Control", "Upregulated")),
    tdp_43_status = recode(
      tdp_43_status,
      `TDP-43 positive` = "Control",
      `TDP-43 negative` = "TDP-43 pathology"
    ) %>%
      factor(levels = c("Control", "TDP-43 pathology"))
  ) %>%
  group_nest(class_family) %>%
  mutate(
    p = map2(
      data, class_family,
      \(x, y) {
        test_res <- editing_index_de_combined_nb_res %>%
          filter(class_family == y)
        test_res_text <- glue_data(
          test_res,
          "<b>{term}</b><br>p = {signif(padj, 2)}, log2FC = {signif(estimate, 2)}"
        ) %>%
          paste(collapse = "<br>")
        n_tes <- n_up_control_tes %>%
          filter(class_family == y)
        n_tes_text <- glue_data(
          n_tes,
          "<b>{upregulated}</b> n = {n}"
        ) %>%
          paste(collapse = "<br>")
        ggplot(
          x,
          aes(
            x = upregulated,
            y = editing_index,
            fill = upregulated
          )
        ) +
          geom_boxplot(outlier.shape = NA) +
          geom_point(alpha = .8, size = 3, color = "black") +
          geom_line(aes(group = patient_sample_id), alpha = .3) +
          scale_fill_manual(
            values = c("Control" = "#6d6868", "Upregulated" = "#c62a2a"),
            guide = "none"
          ) +
          geom_richtext(
            aes(
              label = text
            ),
            data = tibble(tdp_43_status = "Control", text = test_res_text),
            x = -Inf, y = Inf,
            hjust = 0, vjust = 1,
            inherit.aes = FALSE,
            size = 2
          ) +
          geom_richtext(
            aes(
              label = text
            ),
            data = tibble(tdp_43_status = "TDP-43 pathology", text = n_tes_text),
            x = -Inf, y = Inf,
            hjust = 0, vjust = 1,
            inherit.aes = FALSE,
            size = 2
          ) +
          scale_y_continuous(
            expand = expansion(mult = c(0, .2))
          ) +
          facet_wrap(vars(tdp_43_status)) +
          labs(
            title = y,
            x = NULL,
            y = "Edit rate\n(edits per 1000bp of reads)"
          ) +
          theme_bw()
      }
    )
  )

p_combined <- patchwork::wrap_plots(
  ps$p,
  nrow = 3,
  axis_titles = "collect"
)
p_combined
ggsave(
  here("dsrna", "plots", "sprint_analysis", "editing_index_de_tes_by_family.pdf"),
  p_combined, width = 12, height = 8
)
```



```{r}
synStoreMany(
  here("dsrna", "data", "repeat_tdp43_pos_vs_neg_deseq.csv.gz"),
  parentId = "syn64829592",
  forceVersion = FALSE
)
```
